engines/python/setup/djl_python/rolling_batch/neuron_vllm_rolling_batch.py

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python
+#
+# Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file
+# except in compliance with the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS"
+# BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+
+from typing import Optional
+from collections import OrderedDict, defaultdict
+
+from vllm import EngineArgs, LLMEngine, SamplingParams
+from vllm.utils import random_uuid
+from djl_python.rolling_batch.rolling_batch import RollingBatch, stop_on_any_exception, Token, filter_unused_generation_params
+from djl_python.rolling_batch.rolling_batch_vllm_utils import (
+    update_request_cache_with_output, get_lora_request_params, DTYPE_MAPPER,
+    FINISH_REASON_MAPPER)
+from djl_python.properties_manager.tnx_properties import TransformerNeuronXProperties
+
+VLLM_GENERATION_PARAMS = set(SamplingParams().__dict__.keys())
+
+
+class NeuronVLLMRollingBatch(RollingBatch):
+    """
+    NeuronVLLMRollingBatch connects the handler to the backend (VLLM inference). It receives new
+    requests from the handler and sends them to the backend when there is space available in the batch.
+    It also gets any new tokens from the backend and sends them back to the handler.
+    """
+
+    def __init__(self,
+                 config: TransformerNeuronXProperties,
+                 preloaded_model: Optional["PretrainedModel"] = None) -> None:
+        """
+        Initializes the NeuronVLLMRollingBatch.
+
+        :param config: other properties of the model, such as decoder strategy
+        :param preloaded_model: option for providing an already loaded model bypassing vllm model loading
+        """
+
+        self.vllm_configs = config
+        super().__init__(self.vllm_configs)
+
+        args = EngineArgs(
+            model=self.vllm_configs.model_id_or_path,
+            preloaded_model=preloaded_model,
+            tensor_parallel_size=self.vllm_configs.tensor_parallel_degree,
+            dtype=DTYPE_MAPPER[self.vllm_configs.dtype],
+            seed=0,
+            max_model_len=self.vllm_configs.n_positions,
+            max_num_seqs=self.vllm_configs.max_rolling_batch_size,
+            block_size=self.vllm_configs.n_positions,
+            trust_remote_code=self.vllm_configs.trust_remote_code,
+            revision=self.vllm_configs.revision,
+        )
```

Comment:
The only other place where we have deviation is the init parameters. Right now the NeuronVLLMRB class init also can take a preloaded model to bypass vllm model loading, which is slow and does not support advanced neuron compilation configs for better performance. I do think we can merge them but it would also require a change to the vllm rb properties in order to allow for a pretrained model as part of the properties, and neuron specific EngineArgs.

I will take a stab at a refactor and see if there is a way to combine.

Reply:
I see, I think it would be good to explore this for consolidating the two handlers. But that may take some time, and based on Sindhu's comments seems like this refactor effort is already ongoing and might lead to some other changes. I'm fine with this for now as well, given we probably want it in for the release

Reply:
Updated to combine them, testing now and if everything is good I will squash and merge.